### ISSUE TITLE
--help on subcommand not showing options

### ISSUE THREAD

**Issue by blendsdk** [yargs/yargs#1131]
In the following example calling `sayhello --help` does not show the option `--name` in the generated help

```

// This code is in TypeScript

#!/usr/bin/env node
import * as path from 'path';
import * as yargs from 'yargs';

let argv = yargs
    .command('sayhello', 'greet someone', (yargs: yargs.Argv) => {
        return yargs.option('name', {
            type: 'string',
            describe: 'The name to greet',
            demand: true
        })
    }, (args: yargs.Arguments) => {
        console.log(`Hello ${args.name}`);
    })
    .demandCommand()
    .showHelpOnFail(true)
    .argv;
```

When running:
```
script sayhello --help
```

It returns:
```
Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
```

When running:
```
script sayhello 
```

It returns:
```
greet someone

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
  --name     The name to greet                               [string] [required]
```

Why does `script sayhello --help` not work? What am I doing wrong here?

Thank you.

**Comment by aorinevo** [comment]
Hi blendsdk,

Can you try calling showHelpOnFail before command?
```js
let argv = yargs
    .showHelpOnFail(true)
    .command('sayhello', 'greet someone', (yargs: yargs.Argv) => {
        return yargs.option('name', {
            type: 'string',
            describe: 'The name to greet',
            demand: true
        })
    }, (args: yargs.Arguments) => {
        console.log(`Hello ${args.name}`);
    })
    .demandCommand()
    .argv;
```

**Comment by aorinevo** [comment]
Hi alvis,

Can you post the code snippet that sets up your yargs?

**Comment by aorinevo** [comment]
In my app, I noticed that order matters when calling `.help`.  For example, if I call `.help` immediately after requiring yargs, the help menu shows using the `-h` and `--help` flags.  If I call `.help` after `.options`, help menu does not show using `-h`.

```js
// help menu shows with -h and --help flag
const coremetrics = require('../scripts/utils/coremetrics');

module.exports = require('@aorinevo/yargs')
  .help()
  .middleware([coremetrics])
  .commandDir('cmds')
  .options({
    help: {
      alias: 'h',
      describe: 'help',
    },
  }).argv;
```


```js
// help menu shows with --help but not -h flag.
const coremetrics = require('../scripts/utils/coremetrics');

module.exports = require('@aorinevo/yargs')
  .middleware([coremetrics])
  .commandDir('cmds')
  .options({
    help: {
      alias: 'h',
      describe: 'help',
    },
  }).help().argv;
```

**Comment by alvis** [comment]
My config doesn't use `.help()` at all.

It fails with a config as simple as
``` typescript
import * as yargs from 'yargs';
const args = yargs.usage('Usage: ').demandCommand(1).argv;
```

One more interesting thing is if I further simplify the config by removing `.demandCommand()`, even `-h` doesn't work. It prints nothing.

**Comment by lizardruss** [comment]
I ran into this issue today. Here's the workarounds I found:

1. Turn off `esModuleInterop` in your `tsconfig.json`
2. Use `require` instead of `import`

This works:
```typescript
require('yargs').usage('Usage: ').demandCommand(1).argv;
```

This fails:
```typescript
import * as yargs from 'yargs';
yargs.usage('Usage: ').demandCommand(1).argv;
```

I'm guessing that the code generated for imports when `esModuleInterop` is on is causing this. I hope that helps!

**Comment by alvis** [comment]
Can confirm that what @lizardruss described is valid on typescript 2.8.3. It's very weird as esModuleInterop doesn't affect anything, but just --help. Will investigate.

**Comment by alvis** [comment]
So actually it's a known potential issue.
The magic behind esModuleInterop is the __importStar helper as follow
```typescript
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
```

With `var k in mod`, many packages should work just fine as it effectively copies proerties from the original export. However, it's also noted in the [document](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#support-for-import-d-from-cjs-form-commonjs-modules-with---esmoduleinterop) that is may break existing code bases:

> Note: The new behavior is added under a flag to avoid unwarranted breaks to existing code bases. We highly recommend applying it both to new and existing projects. For existing projects, namespace imports (import * as express from "express"; express();) will need to be converted to default imports (import express from "express"; express();).

In other words, the correct way to fix it simply:
```typescript
import yargs from 'yargs';
```

**Comment by bcoe** [comment]
@alvis thanks for digging into this! feel free to open a follow up issue, or better still to submit a patch, if there's anything we can be doing on yargs' end to better interoperate with TypeScript.

**Comment by cspotcode** [comment]
I was hitting the same problem, and the solution works perfectly: stop using a wildcard import.  However, it was not immediately clear *why* it fails:

`__importStar` attempt to copy all enumerable properties off of `yargs` onto the synthetic exports object that it creates.  When it tries to copy `argv`, it triggers yargs's getter, which parses `process.argv`.  This happens at the import statement, before any sort of `yargs.option().command()`... declarations.

yargs could potentially avoid this by making `argv` an inherited property of the exports object, thus avoiding the eager copy in `__importStar` which uses a `hasOwnProperty` check.

**Comment by euberdeveloper** [comment]
> I ran into this issue today. Here's the workarounds I found:
> 
> 1. Turn off `esModuleInterop` in your `tsconfig.json`
> 2. Use `require` instead of `import`
> 
> This works:
> 
> ```ts
> require('yargs').usage('Usage: ').demandCommand(1).argv;
> ```
> 
> This fails:
> 
> ```ts
> import * as yargs from 'yargs';
> yargs.usage('Usage: ').demandCommand(1).argv;
> ```
> 
> I'm guessing that the code generated for imports when `esModuleInterop` is on is causing this. I hope that helps!

Thank you, your answer was useful.

I was using typescript and I noticed that if you use `esModuleInterop` and do not want it to remove it, you can use the default import

```js
import yargs from 'yargs';
```

instead of

```js
import * as yargs from 'yargs';
```
